[seek] replaces constant seek with additive seek support

xhaggi opened the pull request:
This is the follow-up of #5407 and adds additive seek support.

Hitting the seek actions more than once increases or decreases how far it will seek after a delay (currently 1000ms) using these step sizes: +/- 15s, 30s, 60s, 3m, 5m, 10m, 30m.

Each seek step can be enabled/disabled in Settings > Video > Playback > Configuire seek steps.

![screenshot000](https://cloud.githubusercontent.com/assets/1878991/6065039/6fc1d3fa-ad62-11e4-9229-0058fe0426ae.png)

xhaggi committed:
[seek] adds additive seek support

da-anda commented:
mind adding 15s as well? If you only miss a word jumping 30s seems too long for me. I know that there is "small_step_back" but that's hard to map on a 5 button remote. Also with 15s added it will be just like in MediaPortal and ease the move to Kodi for MP users ;)

gurabli commented:
I agree with @da-anda There should be 15s too. If I1m correct, there was a discussion about the possibility that a user can modify add/remove custom values, like in MediaPorrtal? Wouldn't that be even better?

da-anda commented:
@gurabli I had a chat with @xhaggi in IRC and there will be the possibility to select the desired seek steps from within GUI. Custom values could be defined via AdvancedSettings, but should probably not be needed. Making it fully editable from within GUI would be way to complicated to do and IMO is a task for later.

gurabli commented:
@da-anda OK, it sounds excellent, and I completely agree! One of the very few features that are really missing from Kodi, so looking very much to have this option finally!

xhaggi commented:
added some fixes.

BTW currently this only works if you use the following keymaps (or ctrl + left/right, which is default mapped to analogseek keys, but do not work under OSX).
```
<FullscreenVideo>
   <keyboard>
     <left>analogseekback</left>
     <right>analogseekforward</right>
   </keyboard>
</FullscreenVideo>
```

xhaggi commented:
@da-anda mind taking a look at the RFC commit which adds the possibility to configure the seek steps in the GUI.
![screenshot000](https://cloud.githubusercontent.com/assets/1878991/6065039/6fc1d3fa-ad62-11e4-9229-0058fe0426ae.png)

Something to say about how it works. If you select only two steps backwards and forwards (-60s, -30s, 30s, 60s) and you press the seeking key 3 or 4 times the last seeking step (-60s/60s) will be repeated as often as you pressed the key after reaching the last seeking step. You can also select different steps for each direction, for example -10m, -3m, 15s, 30s, 60s, 3m, 10m.

da-anda commented:
@xhaggi looks and sounds good. Just a question - what happens if you only have two seek steps defined for seek forward (30,60), have clicked 5 times forward and then click twice back? Will it substract 120sec or 90sec?

xhaggi commented:
If you press 5 times forward and 2 times backward within the seek time it will end with 5-2 = 3 keystrokes.
And as you pressed the key more times as you have configured seek steps (5-2 = 3) it will end up with the following calculation.

count of keystrokes (5-2 = 3) - size of seek steps (2) + 1 (= 2) * last seek step seconds (60s) = 120s.

xhaggi commented:
> edit: I know it requires some refactoring, but can we use the current actions for this long term? Will you try to refactor that stuff?

what do you mean?

xhaggi committed:
[seek] use SeekHandler for ACTION_STEP_FORWARD and ACTION_STEP_BACK

xhaggi commented:
@FernetMenta mind taking a look too?

xhaggi commented:
@da-anda I've dropped the option to enable/disable additive seeking + the old seeking code in CSeekHandler. Now it use the additive seeking as default implementation. I've also changed the action handling for ACTION_STEP_FORWARD and ACTION_STEP_BACK in CGUIWindowFullScreen to use the SeekHandler instead of the players seek method.

xhaggi commented:
@gurabli it would be nice if you could do some tests ;)

MartijnKaijser commented on language/English/strings.po:
```diff
@@ -16193,9 +16197,14 @@ msgctxt "#37041"
+#. Description of setting "Videos -> Playback -> Configure seek steps"
+#: system/settings/settings.xml
+msgctxt "#37042"
+msgid "Configure seek steps"
```
could be explained a bit better as description i think ;)

xhaggi replied:
> da-anda commented on xbmc/settings/AdvancedSettings.cpp in 11eeaea 3 hours ago 
>this needs to be more descriptive, but I guess you know and it's a placeholder for now
>xhaggi replied 20 minutes ago
yep, it would be nice if you could write down a nice description i could use for it ;)

;)

topfs2 commented:
I'm a tad tired atm so might have missread but just want to echo support for skipping the setting and use this mode over the old? To me it seems better in essentially every way :) I can't really see the usecase were the other isn't covered by this.

topfs2 commented:
And awesome that your working on it, will love to see it in kodi :)

gurabli commented:
@xhaggi Of course, but please let me know in details how can I test:) I`m using Openelec 5.0. What should I do to test?
One remark, I'm leaving now on business trip in few hours time and returning home on Monday, so until then, I!m sorry to say, but I can't do anything:( But if you send me detailed instructions how to add these to OE I would be very happy to test!

xhaggi commented:
@gurabli ah i thought you are able to do it by your-self ;)

gurabli commented:
@xhaggi Sorry, I said it in my first post that I don't know anything about coding or such, but I'm here to help in testing or any other way you might need or find it useful. If you send me an email how can I test it, I will surely do it!

da-anda commented:
@topfs2 with this feature in I'd like to change our default keymaps and remove BIG_STEP_FORWARD/BACK from the up/down keys and replace them with prev/next chapter for videos and prev/next channel for liveTV. And while I also think it's a better way to seek I already hear the complaints that the seeks are not instant and blah blah. But those that want the old behavior can do so by only selecting one step size for SMALL_STEP_FORWARD/BACK and adjust their keymaps to use BIG_STEP_FORWARD/BACK again.
As for the removal of the seek type selector - I agree that we don't need it to differentiate between single step seek and additive seeking because you can simmulate old behavior by only selecting one seek step, but it would be a way to switch between time based seeks and percentual seeks (which we support already and can be enabled via as.xml). Not sure if we still need the percentual seek with additive seeking in place, but if we want to keep it we could add this setting to the GUI.

gurabli commented:
@da-anda People complain because old habits are hard to get over:) Most of them never ever used additive seeking, and they don't even know what it is. Additive seeking IS the seeking, light years more advanced as the old analog way. Of course, the best software is where a user can decide what to do and how. So those who are not happy with additive seeking, just stick to the old outdated type of seeking and nothing changed. For us, who know how important this feature is, it will be a new dimension:) As there is an ON/OFF for this to be set up, this is a MUST for Kodi as well.
I don't see any reason to keep percentual seek with additive seeking, but if it is not hard to have it and leave it, it doesn't complicate things, just leave an option there to turn on or off, and those few people will be also happy who need this:)
Don't listen to those who complain about additive seeking, probably they don`t even know what it is and how it functions in real life:)
OFF: one of the best PVR software ever made, VDR, lacks the most important feature to be the all time best: proper, fully functional timeshifting. Devs don't use it, like it (probably can't do it) so it is not even planned:) Those who use Live TV and VDR do not complain, the same reason, don't even know how and when to use it. Their two cents: I record everything and skip the commercials later, etc. Poor excuse. What about watching live sports? Funny thing, but at least we have MediaPortal's excellent TV Server:) Sorry for the long post.

NedScott commented:
I hesitate to suggest this, because I don't want it to seem like we have to add every conceivable increment, but I think this should include 7 seconds as an optional increment. We've used 7 seconds for years for smallstepback, and it's the perfect amount for "what did he just say?". It's a specific length that we know works.

xhaggi commented:
@NedScott done ;)

da-anda commented:
for anybody interrested - I just triggered a windows testbuild - should be on our mirrors in max 1h from now.

xhaggi commented:
After some hours working on this, several parts are improved. 

First, it's now possible to configure the seek delay in the GUI (500ms to 3000ms, steps 250ms, default: 1000ms).

Second, if you configure only one seek step per direction the seek is applied immediately (without the specified delay) for those who like the old behavior.

Any changes to the seek settings are reflected immediately to the SeekHandler, regardless of whether you are currently playback a movie.

xhaggi committed:
[gui] fix: dialog seek bar slider not updated correctly

xhaggi commented:
@da-anda @ronie I found an issue with the slider which indicates the seek position used in dialog seek bar. After introducing #4739 the slider didn't work anymore. I've fixed it by adding the old implementation. I've also rewritten the whole seek percentage calculation so that the slider works the same for both analog and smart seeking.

@da-anda please create a new test-build, the old one is out-dated

da-anda commented:
thanks, new testbuild is on it's way, should be ready in 1-2h after all the other builds in line have finished.

@ronie - might be nice to see the seek target/progress a bit bigger. In MediaPortal they show a stripped down version of the video OSD with the big progress bar and above the OSD the step size (like "+60 sec") is shown. Can we do something like this as well? It's really hard to see how far you seek in this small popup in the upper right corner atm.

xhaggi commented:
@da-anda @ronie I found the root cause why the new marco `CONTROL_SELECT_ITEM` which is used to set the percentage value of the control do not work for the "slider" in CGUIDialogSeekBar. First, the constant used in the macro was wrong :) and secondly, the macro sends a message which was not delegated to the super class in `CGUIDialogSeekBar::OnMessage()`. I've also renamed the define `POPUP_SEEK_SLIDER` to `POPUP_SEEK_PROGRESS` because the control is not limited to be a slider. It also might be a progressbar control. That's the reason why Jonathan introduced this new message processing to set the percentage value of a "progressbar / slider / ..." control.

xhaggi commented:
jenkins build this please

ronie commented:
tested and the slider in the seek dialog behave as it should now.

@xhaggi since you're looking into this code area, there's another issue i noticed (happens with & without this patch). if you watch a dvd and skip to the next chapter, the seek label always displays -596:31:23

![chapter](https://cloud.githubusercontent.com/assets/687265/6089064/667522b0-ae5d-11e4-963d-62ba8c30bfd6.jpg)

xhaggi commented:
@ronie thanks for your review .. the other issue is not related to this PR, because it is related to chapters skipping and this is handle somewhere else. but if i found some spare time i could take a look.

da-anda commented:
@xhaggi I noticed that the displayed seek steps are most times 1 second of - so it's displaying "29 sec" instead of "30 sec". Will test again with the very very latest version of this PR.

ronie commented:
> might be nice to see the seek target/progress a bit bigger.

@da-anda i'd rather not change the height of the dialog to much.
the dialog is displayed all throughout the skin and would quickly start to overlap other items.

but with some minimal changes, the seek label would be a lot more readable imo.
before/after:
![b](https://cloud.githubusercontent.com/assets/687265/6089758/4c0e1660-ae68-11e4-8c1b-f026ff2d1822.jpg)
![a](https://cloud.githubusercontent.com/assets/687265/6089761/4eab5612-ae68-11e4-9130-79f5bb8bcd71.jpg)

da-anda commented:
@ronie - can't we add a special case handling only for the fullscreen seek? Too bad.

da-anda commented:
@xhaggi - I think something is broken. I have the steps 15,30,60,180,300,600,900... enabled but when I seek 5 steps forward it doesn't seek 5min but 9:44. Also the individual seek steps are never as expected, they show up as 14s, 45s, 104s, 284s, 584s - so something is wrong.

xhaggi commented:
@da-anda thanks for your feedback. I've made a mistake in the calculation of the seek percentage which is now fixed and force pushed. I can also confirm the issue with the displayed seek offset (14s, 28s, 58s etc.)  after a seek, but I'm not sure how to fix it ATM, because it is set by the Player.

MartijnKaijser commented:
could be the delay of updating the label. after you do a 14 seek and it starts playing again then the label is update it's already 1s playing again or something

xhaggi commented:
@MartijnKaijser yep that would explain the deviation.

xhaggi commented:
@da-anda could you please do another test (with a new test-build) and report back if everything works as expected now.

da-anda commented:
@xhaggi @MartijnKaijser no, it's not playback related - the offset even shows when you're paused and perform the seek. Also the offset seems to increase the bigger the jump is. So on small jumps like 15s it's only 1s off, but on 1 min it was only 46s.

Will trigger an uploaded new windows testbuild and check.

da-anda commented:
@xhaggi looks better now - great.

The offset on larger jumps is now also only ~1s - wonder where that comes from as it even happens when paused. But whatever is causing this offset is also causing another issue. I just did some chapter jumps on a movie (while paused) and got stuck at a certain chapter because it didn't seek to the chapter itself but a few seconds before (so same offset issue it seems) and thus subsequent chapter jumps always ended up at the same chapter marker that was only ~2sec ahead. The issue doesn't seem to be related to this PR because I'm having the very same issue in 14.1 release. @FernetMenta  @elupus any ideas?

@Montellese if you have a look at the seek step selector setting you'll notice that due to the ammount of selected options the "preview" of the settings value becomes way to large. Is there a way to define a special label/value that should be used for this preview purpose? (same could be used for subtitle/audio languages, ...)

xhaggi commented:
@da-anda that sounds good ;) 

If there are no objections we should merge it in and to get more feedback from the community.
After that I'll create a cleanup PR which removes the unnecessary advanced settings (m_videoTimeSeekForward, m_videoTimeSeekBackward) + the related parts in the players, we don't need anymore.

jenkins build this please

da-anda commented:
@jjd-uk - mind having a brief look on the added settings labels and descriptions?

FernetMenta commented:
@da-anda 1s off seems not uncommon for non-accurate seeks. some content has recovery points only every couple of seconds.

HitcherUK commented:
> @xhaggi since you're looking into this code area, there's another issue i noticed (happens with & without this patch). if you watch a dvd and skip to the next chapter, the seek label always displays -596:31:23

Also the same for seeking backwards.

xhaggi commented:
@ronie @HitcherUK please file an issue in our bug tracker. Currently I can't reproduce it and it's not related to this PR.
![screenshot002](https://cloud.githubusercontent.com/assets/1878991/6096363/6857935c-af8d-11e4-980a-d26a4e0779ba.png)

da-anda commented:
> @da-anda 1s off seems not uncommon for non-accurate seeks. some content has recovery points only every couple of seconds.
@FernetMenta  I don't have a problem when the player internally jumps to the next B-frame (or whatever is needed), but the jump sized displayed should not IMO still indicate the user triggered step size

@xhaggi can you backport the progress bar fix for Helix so it might get included in a possible 14.2 build?

FernetMenta commented:
@da-anda agree. if this is not the case, there is an issue with this pr.

da-anda commented on system/settings/settings.xml:
```diff
@@ -405,6 +405,32 @@
       <group id="2">
+        <setting id="videoplayer.seeksteps" type="list[integer]" label="13556" help="37042">
+          <level>2</level>
+          <default>-1800,-900,-600,-300,-180,-60,-30,-15,-7,7,15,30,60,180,300,600,900,1800</default>
```
as mentioned on IRC I'd not enable +7 by default - it's a bit low for a seek forward. -7 is fine for jumps back though

xhaggi replied:
sure makes sense

da-anda commented on xbmc/utils/SeekHandler.cpp:
```diff
@@ -19,51 +19,132 @@
-void CSeekHandler::Seek(bool forward, float amount, float duration)
+int CSeekHandler::GetSeekSeconds(bool forward)
 {
+  m_seekStep = m_seekStep + (forward ? 1 : -1);
```
is "+" here correct? shouldn't it be "*" (so multiply in order to negate the value)?

xhaggi replied:
as discussed at IRC, you know you are wrong here ;)

da-anda commented on xbmc/utils/SeekHandler.cpp:
```diff
@@ -19,51 +19,132 @@
+    if ((size_t)m_seekStep <= m_forwardSeekSteps.size())
+      seconds = m_forwardSeekSteps.at(m_seekStep - 1);
+    else
+      seconds = m_forwardSeekSteps.back() * (m_seekStep - m_forwardSeekSteps.size() + 1);
```
a comment might be nice to have here so that others see what this is all about more easily (when exceeding the selected amount of steps repeat/sum up the last step size)

da-anda commented on xbmc/utils/SeekHandler.cpp:
```diff
@@ -19,51 +19,132 @@
+
+    if (!analogSeek)
+    {
+      if ((m_backwardSeekSteps.size() == 1 && forward == false) ||
```
a little comment might not harm here as well (//don't apply a seek delay if only one seek step for given direction is available)

da-anda commented on xbmc/utils/SeekHandler.cpp:
```diff
@@ -79,16 +160,45 @@ bool CSeekHandler::InProgress() const
       g_infoManager.m_performingSeek = true;
+
+      // calculate the seek time
       double time = g_infoManager.GetTotalPlayTime() * m_percent * 0.01;
```
I assume this is where our rounding difference (the 1sec) offset comes from. We probably should differ between analogue (percentual seeks) and our seek steps with fixed time values. I don't see why we should calculate the fixed time steps into percentages and back - we probably should do it the other way around and calculate the resulting seconds for analogue seeks in CSeekHandler::Seek(). So drop m_percent and use a m_seconds instead.

xhaggi replied:
As discussed at IRC what you've said here will not fix the issue with the 1sec offset issue and
is not related to the SeekHandler calculation.

jjd-uk commented:
@da-anda from a quick look the settings labels and descriptions seem fine. 

My only reservation on the language used from a user point of view, is whether Seek would be understandable to someone who is not a geek. The commercial pvr's I've used always call the step forward/backwards function Skip.

xhaggi commented:
@da-anda @jjd-uk i've changed everything you want ;) please take a look at the fixup commit.

da-anda commented:
after the label fix it's good from my side. Fixing the displayed time offset is task for another PR

xhaggi commented:
@da-anda fixed the label + squashed and now ...

jenkins build and merge

NedScott commented:
I started a thread on the forums for feedback on this, if anyone is interested: http://forum.kodi.tv/showthread.php?tid=218024

da-anda commented:
@NedScott there was already a related thread in dev section :) But having one in help&support might not harm as well. See http://forum.kodi.tv/showthread.php?tid=217616

NedScott commented:
Aye, I just didn't want to unleash waves of users into the dev section. However, to my surprise, there has been little activity, so I might just redirect people to the dev thread.

MilhouseVH commented:
Small problem... if you deselect every seek step in Settings -> Video -> Playback, you'll get a crash when seeking.

http://forum.kodi.tv/showthread.php?tid=211501&pid=1933033#pid1933033

xhaggi commented:
@MilhouseVH i've added `<minimumItems>2</minimumItems>` in the settings.xml but it seems that it didn't work as I expected. @Montellese do you know why this don't work?

MilhouseVH commented:
I don't think `<minimumItems>2</minimumItems>` is a workable solution in any event - if the user selects +7 and +15 as the only two skip steps, we're going to hit the same problem we do now (ie. crash, but now only when seeking backwards).

Would it be easier if we had two sklp step lists, one for seeking backwards and one for seeking forwards, with each list requiring a minimum of one item?

By the way, is there any reason why -7 is enabled by default, when this skip duration is already handled by `<smallstepbackseconds>`?

xhaggi commented:
@MilhouseVH sure, but we should handle it in SeekHandler if we don't have a seek step for the selected direction. Split-up the setting is no solution IMO.

> By the way, is there any reason why -7 is enabled by default, when this skip duration is already handled by <smallstepbackseconds>?

we dropped the small step and route this action to the new smart seeking.

xhaggi commented:
@Montellese or can we do such a complex validation (min 1 value needs to be negative+positive) for a setting?

MilhouseVH commented:
> we dropped the small step and route this action to the new smart seeking.
Agh... that's a shame. I see it's now working as you describe, but for anyone used to a TiVo (where this skip-back functionality has been commonplace for donkeys years) this isn't a good move...

MartijnKaijser commented:
so, we are not TiVo and most of us have never used it

MilhouseVH commented:
Your loss.

NedScott commented:
So it's a good feature and we want to keep good features. Several Team Kodi members have used TiVo. I would also like to see smallstepback stay as an option. I thought only the option to configure it had to be dropped for some reason, but the action ID was still there?

MartijnKaijser commented:
what loss?

MilhouseVH commented:
> but the action ID was still there?

It's still there, but now it is based on the adaptive skip values. So if you configure -15 as your first backward seek value, that's going to be your initial skip back value too, followed by -30 or whatever is your next seek value. So multiple stabs of skipback now jumps you further and further back, instead of a constant 7 seconds each time.

I don't think this is desirable behaviour - a constant skipback works better.

> what loss? 

Neve rmind.

xhaggi commented:
@MilhouseVH sorry, but why do not choose -7,-15,-30 everything would work like before except that you go back 15 sec instead of 14 sec if you click twice ;)

MilhouseVH commented:
If I choose -7, -15, -30, if I click skipback I'm back 7 seconds, if I click it again (quickly, without pausing) I'm back 22 seconds. I'm OK with this when seeking using the left/right buttons, but the smallstepback way of seeking doesn't (IMHO) fit so well with this and should have remained non-adaptive - it's smallstepback, after all.

FernetMenta commented:
I think we need some function to terminate further waiting in seek handler, maybe do by pressing OK. the smallstep action could automate this.

xhaggi commented:
> If I choose -7, -15, -30, if I click skipback I'm back 7 seconds, if I click it again (quickly, without pausing) I'm back 22 seconds.

that's not true .. we do not sum up the selected steps. we use each step separately which means 1 = -7 sec, 2 = -15 sec, 3 = -30 sec.

mkortstiege commented:
I can confirm its working as intended. 2x is seeking back 15s and so on.

MilhouseVH commented:
> that's not true .

Yep, you're right - I stand corrected (tested it again, paid more attention this time). I think I'll just grow to love it... fortunately it's rare for me to press skipback more than twice.

jjd-uk commented:
As someone who is used to the instant skips of the old behaviour I was also unsure of this at first, so I did a lot of testing to ensure you could keep near enough the exact same behaviour as before.

I'm now a convert and finding that adjusting Skip Delay to 250ms provides the perfect balance between almost instant skips, so you can press and release press & release to skip -7 or whatever you have as the lowest. But then making a concerted effort to do very rapid presses allows use of the additive function occasional time it might be of use.

If all you want is a instant -7 skip back then turn off Skip Delay, however this will also obviously apply in the positive direction also.

pdf commented:
When I do small skips in video, it's generally to skip intros, and I do it visually, because I can't remember how long every show's intro is.  Same goes if you're trying to skip in content with unflagged commercials.  The way this feature is implemented makes that impossible without disabling smart seek entirely.  I understand that additive steps are probably calculated without seeking for performance, but if seeks were (optionally?) performed with every press, and only the counter was reset after the timeout period, this feature would be a lot more usable for me.  As it was implemented here, it feels unresponsive and haphazard.

So: :arrow_right: seek `n` seconds; :arrow_right: seek `n` seconds; :arrow_right: seek `n`x2 seconds; :arrow_right: seek `n`x4 seconds; wait `timeout` msecs; :arrow_right: seek `n` seconds.

Rather than: :arrow_right:; :arrow_right:; :arrow_right:; :arrow_right: wait `timeout` msecs; seek `n`x8 seconds.
